File: src/FunConcurrency/src/src/CSharpWebCrawler/Memoize.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WebCrawler
{
    public static class Memoization
    {
        public static Func<T, R> Memoize<T, R>(Func<T, R> func) where T : IComparable
        {
            Dictionary<T, R> cache = new Dictionary<T, R>();
            return arg =>
            {
                if (cache.ContainsKey(arg))
                    return cache[arg];
                return (cache[arg] = func(arg));
            };
        }


        // TODO 
        // (1) Implement Thread-safe memoization function
        // (2) Optionally, implement memoization with Lazy behavior
//        public static Func<T, R> MemoizeThreadSafe<T, R>(Func<T, R> func) where T : IComparable
//        {
//            return null;
//        }
       public static Func<T, Task<R>> MemoizeLazyThreadSafe<T, R>(Func<T, Task<R>> func) where T : IComparable
        {
            ConcurrentDictionary<T, Lazy<Task<R>>> cache = new ConcurrentDictionary<T, Lazy<Task<R>>>();
            return arg => cache.GetOrAdd(arg, a => new Lazy<Task<R>>(() => func(a))).Value;
        }

    }
}

File: src/ParallelPatterns/src/Common/Helpers.FSharp/Memoize.fs
namespace Functional

module Helpers =

    open System
    open System.Threading.Tasks
    open System.Collections.Generic
    open System.Collections.Concurrent

    module Memoize =

        let memoize func =
            let table = Dictionary<_,_>()
            fun x ->   if table.ContainsKey(x) then table.[x]
                        else
                            let result = func x
                            table.[x] <- result
                            result
        let memoize2 f =
            let f = (fun (a,b) -> f a b) |> memoize
            fun a b -> f (a,b)

        let memoize3 f =
            let f = (fun (a,b,c) -> f a b c) |> memoize
            fun a b c -> f (a,b,c)
            
        let memoizeThreadSafe (func: 'a -> 'b) =
            let table = ConcurrentDictionary<'a,'b>()
            fun x ->   table.GetOrAdd(x, func)        


        let memoizeWithEnviction cacheTimeSeconds (caller:string) (f: ('a -> 'b)) =
            let cacheTimes = ConcurrentDictionary<string,DateTime>()
            let cache = ConcurrentDictionary<'a, 'b>()    
            fun x ->
                match cacheTimes.TryGetValue caller with
                | true, time when time < DateTime.UtcNow.AddSeconds(-cacheTimeSeconds)
                    -> cache.TryRemove(x) |> ignore
                | _ -> ()
                cache.GetOrAdd(x, fun x -> 
                    cacheTimes.AddOrUpdate(caller, DateTime.UtcNow, fun _ _ ->DateTime.UtcNow)|> ignore
                    f(x)
                    )
    
        let memoizeWithEnvictionAsync cacheTimeSeconds (caller:string) (f: ('a -> Async<'b>)) =
            let cacheTimes = ConcurrentDictionary<string,DateTime>()
            let cache = ConcurrentDictionary<'a, System.Threading.Tasks.Task<'b>>()    
            fun x -> 
                match cacheTimes.TryGetValue caller with
                | true, time when time < DateTime.UtcNow.AddSeconds(-cacheTimeSeconds)
                    -> cache.TryRemove(x) |> ignore
                | _ -> ()
                cache.GetOrAdd(x, fun x -> 
                    cacheTimes.AddOrUpdate(caller, DateTime.UtcNow, fun _ _ ->DateTime.UtcNow)|> ignore
                    f(x) |> Async.StartAsTask
                    ) |> Async.AwaitTask  